lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +321,320 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
   QueuedTasks.push(std::move(T));
 }
 
-bool TaskQueue::execute(TaskBeganCallback Began, TaskFinishedCallback Finished,
-                        TaskSignalledCallback Signalled) {
+// Owns Tasks, handles correspondence with FDs and pids.
+// FIXME: only handles stdout fds.
+class TaskMap {
+  using PidToTaskMap = llvm::DenseMap<pid_t, std::unique_ptr<Task>>;
+  PidToTaskMap TasksByPid;
+
+public:
+  TaskMap() = default;
+
+  bool empty() const { return TasksByPid.empty(); }
+  unsigned size() const { return TasksByPid.size(); }
+
+  void add(std::unique_ptr<Task> T) { TasksByPid[T->getPid()] = std::move(T); }
+
+  Task &findTaskForFd(const int fd) {
+    auto predicate = [&fd](PidToTaskMap::value_type &value) -> bool {
+      return value.second->getPipe() == fd;
+    };
+    auto iter = std::find_if(TasksByPid.begin(), TasksByPid.end(), predicate);
+    assert(iter != TasksByPid.end() &&
+           "All outstanding fds must be associated with a  Task");
+    return *iter->second;
+  }
+
+  void eraseTask(Task &T) { TasksByPid.erase(T.getPid()); }
```

Comment:
If nothing looks these up by PID except for when they're erased, why use a map keyed by PID?

Reply:
It's a fair cop. I thought it might be better to just refactor without making algorithmic changes.
Those could be saved for a subsequent PR. But what do you think?

Reply:
Ah, I should have checked the old code to see what changed and what didn't. You're right—a plain refactor is better than mixing refactoring and behavior changes!

Reply:
OK, tx!